tools/python/google_maps_bookmarks.py

```diff
@@ -2,97 +2,236 @@
             else:
-                self.kml_file = bookmark_list_name + ".kml"
                 break
-        print()
-        self.places = []
+        
+    def get_json(self, url):
+        max_attempts = 3
+        for retry in range(max_attempts):
+            try:
+                response = urllib.request.urlopen(url)
+                return json.load(response)
+            except urllib.error.URLError:
+                print(f"Couldn't connect to Google Maps. Retrying... ({retry + 1}/{max_attempts})")
+                if retry < max_attempts - 1:
+                    continue
+                else:
+                    raise
 
-    def parse_csv(self):
-        with open(self.csv_file, newline='') as file:
-            row_count = sum(1 for _ in file)
-            file.seek(0)
-            csvreader = csv.reader(file, delimiter=',')
-            next(csvreader)  # skip header
-            for idx, row in enumerate(csvreader):
-                name = row[0]
-                description = row[1]
-                url = row[2]
-                print(f"\rProgress: {idx + 1}/{row_count - 1} Parsing {name}...", end='')
-                try:
-                    if url.startswith("https://www.google.com/maps/search/"):
-                        coordinates = url.split('/')[-1].split(',')
-                        coordinates.reverse()
-                        coordinates = ','.join(coordinates)
-                    elif url.startswith('https://www.google.com/maps/place/'):
-                        ftid = url.split('!1s')[-1]
-                        params = {'key': self.api_key, 'fields': 'geometry', 'ftid': ftid}
-                        places_url = "https://maps.googleapis.com/maps/api/place/details/json?" \
-                                     + urllib.parse.urlencode(params)
-                        try:
-                            data = get_json(places_url)
-                            location = data['result']['geometry']['location']
-                            coordinates = ','.join([str(location['lng']), str(location['lat'])])
-                        except (urllib.error.URLError, KeyError):
-                            print(f"Couldn't extract coordinates from Googe Maps. Skipping {name}")
-                            continue
+    def get_coordinates_from_google_api(self, api_key, q=None, cid=None):
+        url = None
+        if q:
+            # Sometimes the 'q' parameter is a comma-separated lat long pair
+            if ',' in q and all(part.replace('.', '', 1).replace('-', '', 1).isdigit() for part in q.split(',')):
+                return q.split(',')
+            else:
+                url = f"https://maps.googleapis.com/maps/api/place/textsearch/json?query={urllib.parse.quote(q)}&key={api_key}"
+                return None
+        elif cid:
+            print(f"This script does not yet support extracting coordinates from a CID. Skipping {cid}" )
+        else:
+            return None
+
+        if url:
```

Comment:
You can call the `get_json` function to get a json from your url. No need to write the code for fetching a json again.